### PR TITLE
[export] Error grad mode op in export API

### PR DESCRIPTION
**Pull request by andrewlee302**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #116527
* __->__ #116339

Summary:
As current export doesn't support training, so grad mode ops doesn't
make sense. To avoid the confusion, we choose to early error if there
exist grad mode ops.

Test Plan:
python test/export/test_safeguard.py

### CODE REVIEW

File: torch/_functorch/aot_autograd.py [hunk 1]
```diff
@@ -1019,7 +1019,8 @@ def fn_to_trace(*args):
     # Next, the input args
     full_args.extend(args)
 
-    with ctx():
+    from torch.export.safeguard import GradModeUnsupportedSafeguard
+    with ctx(), GradModeUnsupportedSafeguard():
```
Comment (andrewlee302) [inline torch/_functorch/aot_autograd.py:1023]: The reason that I don't put the safeguard ctx at the early place (such as https://github.com/pytorch/pytorch/blob/65c5eed01d3bf1b971f97c143beaa9dd8b0be0df/torch/export/_trace.py#L370), because `torch.no_grad` could be possibly invoked here and the safeguard will capture and error, which is not expected.

File: torch/export/_safeguard.py [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+class GradModeUnsupportedSafeguard(TorchFunctionMode):
+    def __torch_function__(self, func, types, args=(), kwargs=None):
+        kwargs = kwargs or {}
+        unsupported_grad_mode_ops = [
+            torch._C._set_grad_enabled,
+        ]
+        if func in unsupported_grad_mode_ops and torch._C._get_dispatch_mode(
+            torch._C._TorchDispatchModeKey.PROXY
+        ):
+            raise RuntimeError(f"{func} is not supported for grad mode")
```
Comment (tugsbayasgalan) [inline torch/export/_safeguard.py:14]: This error message is bit confusing. Can you change it to something like:
```
RuntimeError(f"Encountered autograd state manager op {func} while exporting. This is unsafe because we don't capture this op in torch.export today hence we can't reflect the user intention soundly")

File: test/export/test_safeguard.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+            with torch.set_grad_enabled(True):
+                b = a + a
+            return b
+
+        a = torch.randn(10)
+        with torch.no_grad():
+            export(f1, (a,))
+            export(f3, (a,))
+
+            with self.assertRaises(RuntimeError):
```
Comment (tugsbayasgalan) [inline test/export/test_safeguard.py:34]: Please check the error messages here as well.

File: torch/export/_safeguard.py [hunk 4]
```diff
@@ -0,0 +1,15 @@
+import torch
+from torch.overrides import TorchFunctionMode
+
+
+class GradModeUnsupportedSafeguard(TorchFunctionMode):
+    def __torch_function__(self, func, types, args=(), kwargs=None):
+        kwargs = kwargs or {}
+        unsupported_grad_mode_ops = [
```
Comment (tugsbayasgalan) [inline torch/export/_safeguard.py:8]: If you are going to add the other similar ops in the follow up PR, can you leave a TODO here?

File: torch/export/_safeguard.py [hunk 5]
```diff
@@ -0,0 +1,15 @@
+import torch
+from torch.overrides import TorchFunctionMode
+
+
+class GradModeUnsupportedSafeguard(TorchFunctionMode):
```
Comment (tugsbayasgalan) [inline torch/export/_safeguard.py:5]: Leave a docstring describing what this does?

File: torch/export/_safeguard.py [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+    def __torch_function__(self, func, types, args=(), kwargs=None):
+        kwargs = kwargs or {}
+        unsupported_grad_mode_ops = [
+            torch._C._set_grad_enabled,
+        ]
+        if func in unsupported_grad_mode_ops and torch._C._get_dispatch_mode(
```
Comment (tugsbayasgalan) [inline torch/export/_safeguard.py:11]: Leave comment on how this approach works and why we are checking the mode here? Also cc @bdhirsh

File: torch/export/_safeguard.py (hunk 5 above)
Comment (tugsbayasgalan) [inline torch/export/_safeguard.py:5]: Rename this to: "DetectUnsafeGlobalStateOpsInExportMode"?

File: torch/export/_safeguard.py (hunk 4 above)
Comment (andrewlee302) [reply torch/export/_safeguard.py:8]: I regard the other ops to be added as a new task instead of a follow-up. For now, I don't have ideas about them. You may elaborate later or now what the TODO is so that I can understand what's the relation for this one and future PRs.

File: torch/export/_safeguard.py (hunk 5 above)
Comment (andrewlee302) [reply torch/export/_safeguard.py:5]: Looks like you try to generalize the functionality on global state ops, instead of just grad mode state. Do you intend to do so? `InExportMode` may be redundant, as the file path indicates that.

File: test/onnx/test_fx_to_onnx.py [hunk 7]
```diff
@@ -669,11 +669,12 @@ def forward(self, x):
                 return self.normal.sample(x.shape)
 
         x = torch.randn(2, 3)
-        exported_program = torch.export.export(Model(), args=(x,))
-        _ = torch.onnx.dynamo_export(
-            exported_program,
-            x,
-        )
+        with torch.no_grad():
```
Comment (andrewlee302) [inline test/onnx/test_fx_to_onnx.py:672]: Explicitly add it because the defined Model invokes `troch.no_grad` under the hood and the dynamo produces the graph IR with `torch._C._set_grad_enabled` op which we intend not to allow in the later stage. Setting the grade is false before export can eliminate the `torch._C._set_grad_enabled` op in the graph IR, to make this test case pass.

File: test/export/test_safeguard.py [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+                b = a + a
+            return b
+
+        def f3(a):
+            with torch.set_grad_enabled(False):
+                b = a + a
+            return b
+
+        def f4(a):
+            with torch.set_grad_enabled(True):
```
Comment (tugsbayasgalan) [inline test/export/test_safeguard.py:25]: I learned from Brian that this is unsafe to globally manipulate state like this. Can you wrap the entire test function in a context manager that set the old values back after the test is complete?

File: torch/export/_safeguard.py (hunk 4 above)
Comment (tugsbayasgalan) [reply torch/export/_safeguard.py:16]: We right now don't handle:

- Custom autograd functions
- x.requires_grad_
- x.requires_grad = foo: bool

At least the latter two might be doable within this PR, can you add test cases for the latter two and see if we can intercept it with our new mode?

File: torch/export/_safeguard.py (hunk 6 above)
Comment (tugsbayasgalan) [reply torch/export/_safeguard.py:22]: Can you rephrase it to just say: "We only enable this mode while tracing by checking the torch dispatch mode to see if there is any active ProxyMode"

File: test/onnx/test_fx_to_onnx.py (hunk 7 above)
Comment (tugsbayasgalan) [reply test/onnx/test_fx_to_onnx.py:672]: cc: @thiagocrepaldi

File: torch/export/_safeguard.py (hunk 4 above)
Comment (tugsbayasgalan) [reply torch/export/_safeguard.py:16]: For custom autograd functions, dynamo should make them as special operators, so you can write a test case and see how dynamo handles them. We could probably just error by checking the dynamo produced graph has such special ops. This part can be done as a seperate PR

File: test/export/test_safeguard.py (hunk 8 above)
Comment (andrewlee302) [reply test/export/test_safeguard.py:25]: I think it will revert to the previous status when jumping out of the context. Is what you expect?

File: torch/export/_safeguard.py (hunk 6 above)
Comment (bdhirsh) [reply torch/export/_safeguard.py:22]: @tugsbayasgalan just checking - is the idea that this mode will only be turned on when using aot-dispatch export, and it will be off when using pre-dispatch export?

Comment (tugsbayasgalan) [reply torch/export/_safeguard.py:22]: Yep!

Comment (tugsbayasgalan) [reply torch/export/_safeguard.py:22]: yeah actually good catch! We will need to check if the mode is pre_dispatch or not first.

Comment (tugsbayasgalan) [reply torch/export/_safeguard.py:22]: @andrewlee302 I think you should also check if the ProxyTorchDispatchMode is not pre-dispatch before erroring. I think the object has the field already.

File: test/export/test_safeguard.py (hunk 8 above)
Comment (tugsbayasgalan) [reply test/export/test_safeguard.py:25]: Oh oops sorry missed the context manager part. No need to change

File: torch/export/_safeguard.py (hunk 4 above)
Comment (andrewlee302) [reply torch/export/_safeguard.py:16]: Current dynamo itself has stopped `Tensor.requires_grad_` (https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/variables/tensor.py#L756C1-L759C51) and `Tensor.requires_grad = ` (https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/variables/builtin.py#L1289C1-L1294C22). So we don't need to worry them for now. 

Do you think we should add a test to guarantee they are unsupported from export API side?

Comment (andrewlee302) [reply torch/export/_safeguard.py:16]: What's about **Custom autograd functions**?

Comment (tugsbayasgalan) [reply torch/export/_safeguard.py:16]: We should error when we see a custom autograd function.

Comment (andrewlee302) [reply torch/export/_safeguard.py:16]: Can you give an example? What's the definition of autograd function?

**Review by tugsbayasgalan — APPROVED** [review]
